pdbtools/pdb_chainbows.py

```diff
@@ -0,0 +1,140 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Renames chain identifiers sequentially, based on TER records.
+
+Since HETATM records are not separated by TER records and usually come together
+at the end of the PDB file, this script will erroneously reassign their chain
+identifiers. It is a format limitation. You have been warned!
+
+Usage:
+    python pdb_chainbows.py <pdb file>
+
+Example:
+    python pdb_chainbows.py 1CTF.pdb
+
+This program is part of the `pdb-tools` suite of utilities and should not be
+distributed isolatedly. The `pdb-tools` were created to quickly manipulate PDB
+files using the terminal, and can be used sequentially, with one tool streaming
+data to another. They are based on old FORTRAN77 code that was taking too much
+effort to maintain and compile. RIP.
+"""
+
+import os
+import string
+import sys
+
+__author__ = "Joao Rodrigues"
+__email__ = "j.p.g.l.m.rodrigues@gmail.com"
+
+
+def check_input(args):
+    """Checks whether to read from stdin/file and validates user input/options.
+    """
+
+    # Defaults
+    fh = sys.stdin  # file handle
+
+    if not len(args):
+        # Reading from pipe with default option
+        if sys.stdin.isatty():
+            sys.stderr.write(__doc__)
+            sys.exit(1)
+
+    elif len(args) == 1:
+        if not os.path.isfile(args[0]):
+            emsg = 'ERROR!! File not found or not readable: \'{}\'\n'
+            sys.stderr.write(emsg.format(args[0]))
+            sys.stderr.write(__doc__)
+            sys.exit(1)
+
+        fh = open(args[0], 'r')
+
+    else:  # Whatever ...
+        emsg = 'ERROR!! Script takes 1 argument, not \'{}\'\n'
+        sys.stderr.write(emsg.format(len(args)))
+        sys.stderr.write(__doc__)
+        sys.exit(1)
+
+    return fh
+
+
+def set_chain_sequence(fhandle):
+    """Sets chains sequentially based on existing TER records."""
+
+    chainlist = list(string.ascii_uppercase) + list(string.ascii_lowercase) + \
+        list(range(0, 10))
+    max_chains = len(chainlist)
+
+    issued_warning = False
+
+    curchain = chainlist.pop(0)
+    records = ('ATOM', 'TER', 'ANISOU')
+    for line in fhandle:
+        if line.startswith(records):
+            line = line[:21] + curchain + line[22:]
+
+            if line.startswith('TER'):
+                try:
+                    curchain = chainlist.pop(0)
+                except IndexError:
+                    emsg = 'ERROR!! Structure contains more than {} TER records.\n'
+                    sys.stderr.write(emsg.format(max_chains))
+                    sys.stderr.write(__doc__)
+                    sys.exit(1)
+
+        elif line.startswith('HETATM') and not issued_warning:
```

Comment:
I believe you can avoid the `HETATM` warning with the following implementation.

```python
line = line[:21] + old_new_chain_map[line[21]] + line[22:]
```

Reply:
Yes, for the 30th time :)